scipy/optimize/_root_scalar.py

```diff
@@ -0,0 +1,237 @@
+    options : dict, optional
+        A dictionary of solver options. E.g. `k` or `maxiter`, see
+        :obj:`show_options()` for details.
+
+    Returns
+    -------
+    sol : RootResults
+        The solution represented as a ``RootResults`` object.
+        Important attributes are: ``root`` the solution , ``converged`` a
+        boolean flag indicating if the algorithm exited successfully and
+        ``flag`` which describes the cause of the termination. See
+        `RootResults` for a description of other attributes.
+
+    See also
+    --------
+    show_options : Additional options accepted by the solvers
+
+    Examples
+    --------
+
+    Find the root of a simple cubic
+
+    >>> from scipy import optimize
+    >>> def f(x):
+    ...     return (x**3 - 1)  # only one real root at x = 1
+
+    >>> def fprime(x):
+    ...     return 3*x**2
+
+    brentq takes as input a bracket
+
+    >>> sol = optimize.root_scalar(f, bracket=[0, 3], method='brentq')
+    >>> sol.root, sol.iterations, sol.function_calls
+    (1.0, 10, 11)
+
+    newton starts takes as input a single point and uses the derivative(s)
+
+    >>> sol = optimize.root_scalar(f, x0=0.2, fprime=fprime, method='newton')
+    >>> sol.root, sol.iterations, sol.function_calls
+    (1.0, 11, 22)
+
+    The function can provide the value and derivative in a single call.
+
+    >>> def f_p_pp(x):
+    ...     return (x**3 - 1), 3*x**2, 6*x
+
+    >>> sol = optimize.root_scalar(f_p_pp, x0=0.2, fprime=True, method='newton')
+    >>> sol.root, sol.iterations, sol.function_calls
+    (1.0, 11, 11)
+
+    >>> sol = optimize.root_scalar(f_p_pp, x0=0.2, fprime=True, fprime2=True, method='newton')
+    >>> sol.root, sol.iterations, sol.function_calls
+    (1.0, 6, 6)
+
+
+    """
+    if not isinstance(args, tuple):
+        args = (args,)
+
+    if options is None:
+        options = {}
+
+    # fun also returns the derivative(s)
+    is_memoized = False
+    if fprime2 is not None and not callable(fprime2):
+        if bool(fprime2):
+            f = MemoizeDer(f)
+            is_memoized = True
+            fprime2 = f.fprime2
+            fprime = f.fprime
+        else:
+            fprime2 = None
+    if fprime is not None and not callable(fprime):
+        if bool(fprime):
+            f = MemoizeDer(f)
+            is_memoized = True
+            fprime = f.fprime
+        else:
+            fprime = None
+
+    # respect solver-specific default tolerances - only pass in if actually set
+    kwargs = {}
+    for k in ['xtol', 'rtol', 'maxiter']:
+        v = locals().get(k)
+        if v is not None:
+            kwargs[k] = v
+
+    # Set any solver-specific options
+    if options:
+        kwargs.update(options)
+    # Always want full_output
+    kwargs.update(full_output=True, disp=False)
+
+    # Pick a method if not specified
+    if not method:
+        if bracket:
+            method = 'brentq'
+        elif x0 is not None:
+            method = 'newton'
+        else:
```

Comment:
Can you please add a "Notes" section that explains that if `bracket` is given, but not `method`, then the default is `brentq`, and if no `bracket` but `x0` then it's `newton`? The "Notes" for [`minimize_scalar`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize_scalar.html) are more verbose than this.

Also maybe in a future PR it might be nice to just make the default `brentq` and have `root_scalar` [get a bracket like `Brent` does](../blob/14142ff70d84a6ce74044a27919c850e893648f7/scipy/optimize/optimize.py#L1863-L1869) by calling [`bracket`](blob/v1.1.0/scipy/optimize/optimize.py#L2256)?

Reply:
I'm not sure I want to hard-code the decision for all-time. It's an interesting issue.
I'm thinking ahead to two current /future PRs.
gh-8876 adds TOMS 748 which may be judged a better default for bracketed toot-finding.
Sidi's method would expand to cover complex-valued functions in the complex domain when given two points.
So other default would be "use the best method available for the situation presented".
Thoughts?

Reply:
> So other default would be "use the best method available for the situation presented".

this makes sense to me

Reply:
Personally I would steer users away from Newton unless `fprime` is given. If `x0` is given but not `fprime`, then I would suggest growing a bracket around `x0` and calling Brent. Secant is not reliable IMO. For example try to solve the single diode model with a bad guess and you get GIGO.

```python
IL, I0, RS, RSH, GAMMA, VT = 7., 1.e-6, 0.01, 20., 1.23, 0.026

def sdm(v, i, il, i0, rs, rsh, gamma,vt):
    vd=v+I*rs
    f = il-i0*(np.exp(vd/gamma/vt)-1)-vd/rsh-i
    return f

VOC=GAMMA*np.log(IL/I0+1.)
I=3.  # current [A]
ARGS=(I, IL, I0, RS, RSH, GAMMA, VT)
V0 = 0.3. # voltage [V] bad guess

root_scalar(sdm, bracket=(0,VOC))

# works!

root_scalar(sdm, x0=VOC)

# also works!

root_scalar(sdm, x0=V0). # bad guess

# fails!
```